compio-runtime/src/attacher.rs

```diff
@@ -53,39 +57,169 @@ impl Attacher {
         }
     }
 
-    /// Check if [`attach`] has been called.
-    pub fn is_attached(&self) -> bool {
+    /// Attach the inner source and get the reference.
+    pub fn try_get(&self) -> io::Result<&S> {
+        self.attach()?;
+        Ok(&self.source)
+    }
+
+    /// Get the reference of the inner source without attaching it.
+    ///
+    /// # Safety
+    ///
+    /// The caller should ensure it is attached before submit an operation with
+    /// it.
+    pub unsafe fn get_unchecked(&self) -> &S {
+        &self.source
+    }
+
+    /// Attach the inner source and get the mutable reference.
+    pub fn try_get_mut(&mut self) -> io::Result<&mut S> {
+        self.attach()?;
+        Ok(&mut self.source)
+    }
+
+    /// Get the mutable reference of the inner source without attaching it.
+    ///
+    /// # Safety
+    ///
+    /// The caller should ensure it is attached before submit an operation with
+    /// it.
+    pub unsafe fn get_unchecked_mut(&mut self) -> &mut S {
+        &mut self.source
+    }
+}
+
+impl<S> Attachable for Attacher<S> {
+    fn is_attached(&self) -> bool {
         self.once.get().is_some()
     }
+}
 
+impl<S: IntoRawFd> IntoRawFd for Attacher<S> {
+    fn into_raw_fd(self) -> RawFd {
+        self.source.into_raw_fd()
+    }
+}
+
+impl<S: FromRawFd> FromRawFd for Attacher<S> {
+    unsafe fn from_raw_fd(fd: RawFd) -> Self {
+        Self::new(S::from_raw_fd(fd))
+    }
+}
+
+impl<S: TryClone + AsRawFd> TryClone for Attacher<S> {
     /// Try clone self with the cloned source. The attach state will be
     /// reserved.
     ///
     /// ## Platform specific
     /// * io-uring/polling: it will try to attach in the current thread if
     ///   needed.
-    pub fn try_clone(&self, source: &impl AsRawFd) -> io::Result<Self> {
-        if cfg!(windows) {
-            Ok(self.clone())
+    fn try_clone(&self) -> io::Result<Self> {
+        let source = self.source.try_clone()?;
+        let new_self = if cfg!(windows) {
+            Self {
+                source,
+                once: self.once.clone(),
+            }
         } else {
-            let new_self = Self::new();
+            let new_self = Self::new(source);
             if self.is_attached() {
-                new_self.attach(source)?;
+                new_self.attach()?;
             }
-            Ok(new_self)
-        }
+            new_self
+        };
+        Ok(new_self)
+    }
+}
+
+impl<S> IntoInner for Attacher<S> {
+    type Inner = S;
+
+    fn into_inner(self) -> Self::Inner {
+        self.source
     }
 }
 
 /// Represents an attachable resource to driver.
 pub trait Attachable {
-    /// Attach self to the global driver.
-    fn attach(&self) -> io::Result<()>;
-
     /// Check if [`Attachable::attach`] has been called.
     fn is_attached(&self) -> bool;
 }
 
+/// Duplicatable file or socket.
+pub trait TryClone: Sized {
```

Comment:
Do we need `TryClone` and `TryAsRawFd` being seeled? i.e. should we allow user to provide their own implementation at the cost of stability guarantee?

Reply:
They should not be hidden, because the users may use them, especially `try_clone`.

I'll mark them sealed.

Reply:
Oh, no. They could not be sealed because there are many other types in other crates implementing them.

Reply:
I think it's OK for the users to implement these traits to implement their own IO source. It won't break our design.